linux/keyman-config/keyman_config/gnome_keyboards_util.py

```diff
@@ -0,0 +1,76 @@
+#!/usr/bin/python3
+import logging
+import os
+from gi.repository import Gio
+
+from gi.overrides.GLib import Variant
+
+
+class GnomeKeyboardsUtil():
+    def __init__(self):
+        self.input_sources = Gio.Settings.new("org.gnome.desktop.input-sources")
+
+    def read_input_sources(self):
+        sourcesVal = self.input_sources.get_value("sources")
+        sources = self._convert_variant_to_array(sourcesVal)
+        return sources
+
+    def write_input_sources(self, sources):
+        sourcesVal = self._convert_array_to_variant(sources)
+        self.input_sources.set_value("sources", sourcesVal)
+
+    def _convert_variant_to_array(self, variant):
+        if variant is None:
+            return []
+
+        values = []
+        # a(ss)
+        nChildren = variant.n_children()
+        for i in range(nChildren):
+            # (ss)
+            val = variant.get_child_value(i)
+            typeVariant = val.get_child_value(0)
+            type = typeVariant.get_string()
+            idVariant = val.get_child_value(1)
+            id = idVariant.get_string()
+            values.append((type, id))
+        return values
+
+    def _convert_array_to_variant(self, array):
+        if len(array) == 0:
+            return None
+
+        children = []
+        for (type, id) in array:
+            typeVariant = Variant.new_string(type)
+            idVariant = Variant.new_string(id)
+            child = Variant.new_tuple(typeVariant, idVariant)
+            children.append(child)
+        return Variant.new_array(None, children)
+
+
+__is_gnome_shell = None
+
+
+def is_gnome_shell():
+    global __is_gnome_shell
+
+    if __is_gnome_shell is None:
+        code = os.system('pidof gnome-shell >/dev/null 2>&1')
+        __is_gnome_shell = (code == 0)
+    return __is_gnome_shell
+
+
+def _reset_gnome_shell():
+    # used in unit tests
+    global __is_gnome_shell
+
+    __is_gnome_shell = None
+
+
+def get_keyboard_id(keyboard, packageDir, ignore_language=False):
+    kmx_file = os.path.join(packageDir, keyboard['id'] + ".kmx")
+    if not ignore_language and "languages" in keyboard and len(keyboard["languages"]) > 0:
+        logging.debug(keyboard["languages"][0])
+        return "%s:%s" % (keyboard["languages"][0]['id'], kmx_file)
```

Comment:
So this always returns the first language in the keyboard metadata? This seems to be an obvious area for future enhancement

Reply:
Definitely. To the user this shows up as a confusing display of the keyboard in the Gnome UI. I'll create an issue. At the moment we can't improve this unless we know the language the user searched for (cf #1456)

Reply:
We're going to improve the install workflow -- in the package install dialog, we should show a selector for each keyboard to choose from the list of languages in the package; if the default can be prepulated from the search per #1456, then that's ideal, but otherwise we still get a better answer.